server/drivers/futaba.c

```diff
@@ -0,0 +1,619 @@
+	{
+		report(RPT_DEBUG, "%s: Kernel driver found. Detatching it...", drvthis->name);
+
+		if( libusb_detach_kernel_driver(my_driver->my_handle,0) )
+		{
+			report(RPT_DEBUG, "%s: Can't detach kernel driver", drvthis->name);
+			retVal--;
+		}
+	}
+	return retVal;
+}
+
+/**
+ * Internal routine to set up usb interface and reset display.
+ * \param drvthis  Pointer to driver structure.
+ * \return         0 on success.
+ */
+int futaba_start_driver(Driver *drvthis)
+{
+	int n;
+	futabaDriver_t *my_driver = &g_futabaDriver;
+	PrivateData *p = drvthis->private_data;
+
+	if( futabaInitDriver(my_driver,drvthis) )
+	{
+		report(RPT_DEBUG, "%s: Failed to init driver", drvthis->name);
+		futaba_shutdown(drvthis);
+	}
+	else
+	{
+		report(RPT_DEBUG, "%s: RESET Device", drvthis->name);
+		if( (n = libusb_reset_device(my_driver->my_handle)) )
+		{
+			report(RPT_DEBUG, "%s: RESET Failed", drvthis->name);
+			libusb_close(my_driver->my_handle);
+			if( futabaInitDriver(my_driver, drvthis) )
+			{
+				report(RPT_DEBUG, "%s: Failed to re-init driver", drvthis->name);
+				futaba_shutdown(drvthis);
+			}
+		}
+	}
+
+	// Claim the interface for us
+	if( (n = libusb_claim_interface(my_driver->my_handle,0)) )
+	{
+		report(RPT_DEBUG, "%s: Failed to claim interface", drvthis->name);
+		libusb_close(my_driver->my_handle);
+		libusb_exit(my_driver->ctx);
+		exit(-1);
+	}
+	else
+	{ // Write the usb handle and context to the private_data
+		p->my_handle = my_driver->my_handle;
+		p->ctx = my_driver->ctx;
+	}
+	return 0;
+}
+
+/**
+ * Internal function to cleanly exit.
+ * \param drvthis  Pointer to driver structure.
+ */
+void futaba_shutdown(Driver *drvthis)
+{
+	PrivateData *p = drvthis->private_data;
+
+	libusb_release_interface(p->my_handle,0);
+	libusb_close(p->my_handle);
+
+	libusb_exit(p->ctx);
+	// printf("Futaba display driver shutdown complete\n");
+}
+
+ /**
+ * Initalize the private data and call futaba_start_driver.
+ * \param drvthis  Pointer to driver structure.
+ * \return         0 on success, -1 on error.
+ */
+
+MODULE_EXPORT int
+futaba_init (Driver *drvthis)
+{
+	PrivateData *p;
+
+	/* Allocate and store private data */
+	p = (PrivateData *) calloc(1, sizeof(PrivateData));
+	if (p == NULL)
+		return -1;
+	if (drvthis->store_private_ptr(drvthis, p))
+		return -1;
+
+	/* initialize private data */
+
+	p->width = 7;
+	p->height = 1;
+        p->old_icon_map = 0;
+	p->framebuf = (char *)malloc(p->width * p->height);
+
+	if (p->framebuf == NULL) {
+		report(RPT_ERR, "%s: unable to create framebuffer", drvthis->name);
+		return -1;
+	}
+
+	p->old_framebuf = (char *)malloc(p->width * p->height);
+	if (p->old_framebuf == NULL) {
+		report(RPT_ERR, "%s: unable to create old_framebuffer", drvthis->name);
+		return -1;
+	}
+
+	/* write to the file */
+	futaba_start_driver(drvthis);
+	memset(p->framebuf, ' ', p->width * p->height);
+	memcpy(p->old_framebuf, p->framebuf,p->width * p->height);
+	report(RPT_DEBUG, "%s: init() done", drvthis->name);
```

Comment:
Use debug() here, so this gets optimized away on non-debug builds ...
The same is true at more places.